RavenDB-10374 hook up to unhandledException and unobservedTaskException

gregolsky opened the pull request.

gregolsky committed:
RavenDB-10374 hook up to unhandledException and unobservedTaskException

ayende commented on src/Raven.Server/Utils/UnhandledExceptions.cs:
```diff
@@ -0,0 +1,28 @@
+            AppDomain.CurrentDomain.UnhandledException += (sender, eventArgs) =>
+            {
+                var exceptionString = $"UnhandledException: { eventArgs.ExceptionObject.ToString() }.";
+                if (logger.IsInfoEnabled)
```
Operations, this is killing us

ppekrol replied:
Done

ayende reviewed (commented):
What about the actual crash? When we have a finalizer error, for example, do we actually log?
Note that our logger is async, so we might not log it in time.

We need something that would flush a _real_ error and wait for it to be written.

ayende commented on src/Raven.Server/Utils/UnhandledExceptions.cs:
```diff
@@ -0,0 +1,28 @@
+                    logger.Info(exceptionString);
+            };
+
+            TaskScheduler.UnobservedTaskException += (sender, eventArgs) =>
```
This is okay as info, but note that we need to do the check _first_

ppekrol committed:
RavenDB-10366
- added the ability to wait for the log to be actually written that will also force flushing (important for unobserved exceptions)
- when we are adding new 'thread state' in logger, we need to set MRE to 'set' to start the logging process again
- unobserved exceptions should be logged at Operations level

ayende commented on src/Sparrow/Logging/Logger.cs:
```diff
@@ -8,29 +9,46 @@ public class Logger
         }
 
-        public void Info(string msg, Exception ex = null)
+        public Task Info(string msg, Exception ex = null, bool withWait = false)
```
Better to have a totally separate overload here.

ayende replied:
Call InfoAsync?

ayende commented on src/Raven.Server/Utils/UnhandledExceptions.cs:
```diff
@@ -0,0 +1,44 @@
+        public static void Track(Logger logger)
+        {
+            AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
+            {
```
This is going to kill the process, it ALWAYS interests us.
That said, are we sure that this actually give us a good error _when running as a service_?

ppekrol replied:
We are re-checking that, just in case.

ayende commented on src/Raven.Server/Utils/UnhandledExceptions.cs:
```diff
@@ -0,0 +1,44 @@
+                    return;
+
+                var exception = args.Exception.ExtractSingleInnerException();
+                if (logger.IsInfoEnabled)
```
already checked

ppekrol committed:
RavenDB-10366 refactoring

ppekrol commented:
Applied fixes for comments.

ayende commented:
didn't push them?

ppekrol commented:
Amended last commit.
